menpofit/dpm/base.py

```diff
@@ -0,0 +1,1062 @@
+            start = time.time()
+            model = self._train(model, pos, kneg, 1)
+            fp = os.path.join(pickle_dev, 'mix.pkl')
+            mio.export_pickle(model, fp)
+            stop = time.time()
+            fp = os.path.join(pickle_dev, 'mix_time.pkl')
+            mio.export_pickle(stop-start, fp)
+
+        try:
+            fp = os.path.join(pickle_dev, 'final.pkl')
+            model = mio.import_pickle(fp)
+        except ValueError:
+            start = time.time()
+            model = self._train(model, pos, neg, 2)
+            fp = os.path.join(pickle_dev, 'final.pkl')
+            mio.export_pickle(model, fp)
+            stop = time.time()
+            fp = os.path.join(pickle_dev, 'final.pkl')
+            mio.export_pickle(stop-start, fp)
+
+    def _split(self, pos_images):
+        # Each component contains different number of parts.
+        # Split the positive exampls's parts according to its part pools.
+        conf = self.config
+        split_pos = []
+        for i in range(conf['partpoolsize']):
+            split_pos.append([])
+        for p in pos_images:
+            part_ids_inpool = conf['mixture_poolid'][p['gmixid']]
+            for i, k in enumerate(part_ids_inpool):
+                s = dict()
+                s['im'] = p['im'][:]
+                s['gmixid'] = 0  # Each independent part is assume to belong to 0th component.
+                s['box_y1'] = [p['box_y1'][i]]
+                s['box_x1'] = [p['box_x1'][i]]
+                s['box_y2'] = [p['box_y2'][i]]
+                s['box_x2'] = [p['box_x2'][i]]
+                split_pos[k].append(dict(s))
+        return split_pos
+
+    def _ln2box(self, pos):
+        # Converts the points into bounding boxes.
+        for i, p in enumerate(pos):
+            parents_lns = self._get_parents_lns(p['gmixid'])
+            lengths = np.linalg.norm(abs(p['pts'][1:] - p['pts'][parents_lns[1:]]), axis=1)
+            boxlen = np.percentile(lengths, 80, interpolation='midpoint')
+            assert(boxlen > 3)  # ensure that boxes are 'big' enough.
+            _t = np.clip(p['pts'] - 1 - boxlen/2, 0, np.inf) # -1 for matlab indexes
+            p['box_x1'] = np.copy(_t[:, 0])
+            p['box_y1'] = np.copy(_t[:, 1])
+            _t = p['pts'] - 1 + boxlen/2  # no check for boundary, -1 for matlab indexes
+            p['box_x2'] = np.copy(_t[:, 0])
+            p['box_y2'] = np.copy(_t[:, 1])
+        return pos
+
+    def _init_model(self, pos_, sbin):
+        areas = np.empty((len(pos_),))
+        for i, el in enumerate(pos_):
+            areas[i] = (el['box_x2'][0] - el['box_x1'][0] + 1) * (el['box_y2'][0] - el['box_y1'][0] + 1)
+        areas = np.sort(areas)
+        area = areas[np.floor(areas.size * 0.2)]  # Pick the 20th percentile area
+        nw = np.sqrt(area)
+
+        im = hog(Image(np.zeros((1, 30, 30))), mode='sparse', algorithm='zhuramanan')  # Calculating HOG features size
```

Comment:
Assuming we want the user to be able to choose features we would need to make this more interactive.